Organizes output of `bin/cake routes` into columns.

cnizzardini opened the pull request:
The motivation behind this PR is cleaning up the output of `bin/cake routes` so data is easier to read for developers maintaining applications with a large number of routes. This change:

- Sorts alphabetically by route name (edit: optionally via `--sort`)
- Previous JSON display is maintained (edit: with the existing) `--verbose` switch
- Updates unit test

Sample output:

```
> bin/cake routes
+-------------------------------------+-------------------------------------------+-------------+-------------------+-------------+-----------+---------------+
| Route name                          | URI template                              | Controller  | Action            | Plugin      | Prefix    | Method(s)     |
+-------------------------------------+-------------------------------------------+-------------+-------------------+-------------+-----------+---------------+
| _controller:_action                 | /{controller}/{action}/*                  |             | index             |             |           |               |
| actors:edit                         | /actors/:id                               | Actors      | edit              |             |           | PUT, PATCH    |
| admin:setup._controller:_action     | /admin/setup/{controller}/{action}/*      |             | index             | Setup       | Admin     |               |
```

Adding the `--verbose` or `-v` switch simply appends a column called Defaults to the end of the rows.

I had considered adding some additional switches such as `--plugin` for limiting to a specific plugin and `--no-plugins` for hiding all plugins, but I think these use-cases can be accomplished with `grep`. This keeps the change relatively small and eliminates the need for additional unit tests.

cnizzardini committed:
Organizes output `bin/cake routes` into columns.

- Previous JSON display is maintained with a --verbose switch
- Updates unit test

cnizzardini committed:
Stickler: Multi-line arrays must have a trailing comma after the last
element.

cnizzardini commented:
If someone wants to point me in the right direction I can add coverage for this line: https://github.com/cakephp/cakephp/pull/15446/checks?check_run_id=2317637645

ADmad commented:
> Sorts alphabetically by route name

I don't like this. Currently the routes are listed in the order in which they are checked in `RoutesCollection::parseRequest()`. That's more useful when debugging route matching issue.

ndm2 commented:
The sorting of [`Router::routes()` was just recently updated](https://github.com/cakephp/cakephp/pull/15248) so that it matches the order in which routes are actually being matched, which is not necessarily identical with the order in which they are connected, hence that specific order can help a lot to figure out problems with non-matching routes, hence personally I'd like to not see this change.

How about introducing an option for different sort orders?

cnizzardini committed:
Adds `sort` switch to optionally sort route names A-Z, default is none

cnizzardini commented:
I am fine with that. I added a `--sort` switch and there is no sorting by default now. The unit test uses `-s` for coverage purposes.

garas commented on src/Command/RoutesCommand.php:
```diff
@@ -57,7 +88,16 @@ public function execute(Arguments $args, ConsoleIo $io): ?int
+            ])
+            ->addOption('sort', [
+                'help' => 'Sorts alphabetically by route name A-Z',
+                'short' => 's',
```
```suggestion
                'short' => 's',
                'boolean' => true,
```

Then `$args->getOption('sort')` will be `true` or `false`

cnizzardini replied:
What's the downside of a simple null check?

garas replied:
Booleans doesn't have value.

Option now consumes next argument as its value

`bin/cake routes --sort --verbose` will set `sort=--verbose`...

garas commented on src/Command/RoutesCommand.php:
```diff
@@ -57,7 +88,16 @@ public function execute(Arguments $args, ConsoleIo $io): ?int
-        $parser->setDescription('Get the list of routes connected in this application.');
+        $parser
+            ->setDescription('Get the list of routes connected in this application.')
+            ->addOption('verbose', [
```
`verbose`/`quite` options already exist by default

cnizzardini committed:
`--verbose` is superfluous, the option already exists

cnizzardini committed:
Use boolean for switch conditions sort and verbose

garas commented on src/Command/RoutesCommand.php:
```diff
@@ -35,14 +35,45 @@ class RoutesCommand extends Command
+                $methods,
+            ];
+
+            if ($args->getOption('verbose') === true) {
```
```suggestion
            if ($args->getOption('verbose')) {
```

cnizzardini committed:
Update src/Command/RoutesCommand.php

remove unnecessary strict type check

Co-authored-by: Edgaras Janušauskas <edgaras.janusauskas@gmail.com>

cnizzardini committed:
remove unnecessary strict type check

cnizzardini committed:
Merge branch '4.next-routes-command' of github.com:cnizzardini/cakephp into 4.next-routes-command

cnizzardini committed:
check for true on verbose switch

ADmad commented on src/Command/RoutesCommand.php:
```diff
@@ -35,14 +35,45 @@ class RoutesCommand extends Command
-        $output = [
-            ['Route name', 'URI template', 'Defaults'],
-        ];
+        $header = ['Route name', 'URI template', 'Controller', 'Action', 'Plugin', 'Prefix', 'Method(s)'];
```
Having `Prefix` and `Plugin` before `Controller` would be more inline with how URL parts are generally mapped.

garas replied:
More exact order is `Plugin` before `Prefix`.

ADmad commented on tests/TestCase/Command/RoutesCommandTest.php:
```diff
@@ -71,26 +71,46 @@ public function testRouteListHelp()
     public function testRouteList()
     {
-        $this->exec('routes');
+        $this->exec('routes -v -s');
```
There should be test for `routes` too without any additional options.

cnizzardini replied:
I think this meets near full coverage as-is. I could add a smaller test for just `-v -s` to test order and the additional column if you think that's better.

cnizzardini replied:
@ADmad Is there an easy way to test the order of the output when using `--sort`?

cnizzardini replied:
Cuz you ain't gonna like what I'm cooking up buddy.

cnizzardini replied:
I just pulled something out of the oven for you buddy. Bon Appétit!

ADmad commented on src/Command/RoutesCommand.php:
```diff
@@ -35,14 +35,45 @@ class RoutesCommand extends Command
-            ksort($route->defaults);
-            $output[] = [$name, $route->template, json_encode($route->defaults)];
+            $methods = '';
+            if (isset($route->defaults['_method']) && is_array($route->defaults['_method'])) {
```
You aren't showing the method when it's a string?

cnizzardini committed:
Change order of columns, put Plugin and Prefix before Controller

cnizzardini committed:
Fix bug, was not showing _method when type string

cnizzardini committed:
Split tests out for `routes -v` and `-s` into separate tests...

Adds ConsoleIntegrationTestTrait::getMessages() to get console
messages as an array to test ordering.

ADmad commented on src/TestSuite/ConsoleIntegrationTestTrait.php:
```diff
@@ -144,6 +144,16 @@ public function useCommandRunner(): void
+     *
+     * @return array
+     */
+    public function getMessages(): array
```
If we are going to add this method then better to name is `getOutMessage()`.

cnizzardini replied:
Would it be better just to provide a getter for `\Cake\TestSuite\Stub\ConsoleOutput`?

ADmad replied:
Well we could just use the protected property `$this->_out` directly too.

Whether we need a public interface for the ConsoleOutput instance can be discussed in a separate issue/PR.

cnizzardini replied:
No, getMessages() is pointless. Not sure why I thought `_out` was private. I removed it and just accessed `_out` directly.

cnizzardini committed:
Removes unnecessary getMessages() method

markstory commented:
Great job @cnizzardini :clap:
